src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -0,0 +1,256 @@
+                        throw new NotSupportedException(); // TODO
+                    case KnownHeaders.ContentType:
+                        if (initialState != CborReaderState.TextString &&
+                            initialState != CborReaderState.UnsignedInteger)
+                        {
+                            throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                        }
+                        reader?.SkipValue();
+                        break;
+                    case KnownHeaders.Kid:
+                    case KnownHeaders.IV:
+                    case KnownHeaders.PartialIV:
+                        if (initialState != CborReaderState.ByteString)
+                        {
+                            throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                        }
+                        reader?.SkipValue();
+                        break;
+                    default:
+                        reader?.SkipValue();
+                        break;
+                }
+            }
+        }
+
+        internal byte[] Encode(bool mustReturnEmptyBstrIfEmpty = false)
+        {
+            if (_headerParameters.Count == 0 && mustReturnEmptyBstrIfEmpty)
+            {
+                return s_emptyBstrEncoded;
+            }
+
+            var writer = new CborWriter();
+            writer.WriteStartMap(_headerParameters.Count);
+            foreach ((CoseHeaderLabel Label, ReadOnlyMemory<byte> EncodedValue) header in this)
+            {
+                CoseHeaderLabel label = header.Label;
+                if (label.LabelAsInt32 != null)
+                {
+                    writer.WriteInt32(label.LabelAsInt32.Value);
+                }
+                else
+                {
+                    Debug.Assert(label.LabelAsString != null);
+                    writer.WriteTextString(label.LabelAsString);
+                }
+                writer.WriteEncodedValue(header.EncodedValue.Span);
+            }
+            writer.WriteEndMap();
+            return writer.Encode();
+        }
+
+        public IEnumerator<(CoseHeaderLabel, ReadOnlyMemory<byte>)> GetEnumerator() => new Enumerator(_headerParameters.GetEnumerator());
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        internal struct Enumerator : IEnumerator<(CoseHeaderLabel, ReadOnlyMemory<byte>)>
+        {
+            private Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>>.Enumerator _dictionaryEnumerator;
+
+            internal Enumerator(Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>>.Enumerator dictionaryEnumerator)
+            {
+                _dictionaryEnumerator = dictionaryEnumerator;
+            }
+
+            public (CoseHeaderLabel, ReadOnlyMemory<byte>) Current => (_dictionaryEnumerator.Current.Key, _dictionaryEnumerator.Current.Value);
```

Comment:
Assuming we go with the "public struct" plan, there should be a `readonly` here somewhere.  Combined with "name the tuple elements" that's

```suggestion
            public readonly (CoseHeaderLabel Label, ReadOnlyMemory<byte> Value) Current => (_dictionaryEnumerator.Current.Key, _dictionaryEnumerator.Current.Value);
```

I think.